icicle-compiler/test/Icicle/Test/Runtime/Array.hs

```diff
@@ -1,96 +1,359 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE TemplateHaskell #-}
 module Icicle.Test.Runtime.Array where
 
+import           Anemone.Foreign.Mempool (Mempool)
 import qualified Anemone.Foreign.Mempool as Mempool
 
-import           Control.Monad.Catch (bracket)
+import           Control.Monad.IO.Class (MonadIO(..))
+import           Control.Monad.Morph (hoist)
+import           Control.Monad.Reader.Class (MonadReader(..))
+import           Control.Monad.Trans.Reader (runReaderT)
+import           Control.Monad.Trans.Resource (MonadResource, allocate)
 
 import qualified Data.List as List
+import           Data.Map (Map)
+import qualified Data.Map.Strict as Map
 
-import           Disorder.Jack
-import           Disorder.Core.IO (testIO)
+import           Hedgehog
+import qualified Hedgehog.Gen as Gen
+import qualified Hedgehog.Range as Range
 
-import           Icicle.Runtime.Array (ArrayLength(..))
+import           Icicle.Runtime.Array (Array, ArrayLength(..), ArrayIndex(..))
 import qualified Icicle.Runtime.Array as Array
-import           Icicle.Test.Arbitrary.Run
 
 import           P
 
 import           System.IO (IO)
 
-import           X.Control.Monad.Trans.Either (EitherT, runEitherT)
 
-
-runEitherIO :: Show x => EitherT x IO a -> IO a
-runEitherIO io = do
-  e <- runEitherT io
-  case e of
-    Left err ->
-      fail (show err)
-    Right x ->
-      pure x
+newMempool :: MonadResource m => m Mempool
+newMempool =
+  snd <$> allocate Mempool.create Mempool.free
 
 prop_capacity :: Property
 prop_capacity =
-  gamble (choose (2, 62) :: Jack Int64) $ \ n ->
-     2 ^ n === Array.calculateCapacity (ArrayLength (2 ^ n))
+  property $ do
+    n <- forAll $ Gen.int (Range.constant 2 62)
+    2 ^ n === Array.calculateCapacity (ArrayLength (2 ^ n))
 
 prop_roundtrip_writes :: Property
 prop_roundtrip_writes =
-  gamble (listOf (bounded :: Jack Int64)) $ \xs0 ->
-  testIO . bracket Mempool.create Mempool.free $ \pool -> do
-    array <- Array.new pool (fromIntegral (length xs0))
+  property $ do
+    xs0 <- forAll . Gen.list (Range.linear 0 100) $ Gen.int64 Range.linearBounded
```

Comment:
so this is a list that grows linearly from 0 to 100, containing int64s of from minBound..maxBound?
💯 💯 💯